File: apps/web/src/lib/tree.ts
import type { TreeNode } from "./api";

export function findFirstFile(nodes: TreeNode[]): TreeNode | null {
  for (const node of nodes) {
    if (node.type === "file") {
      return node;
    }
    const nested = findFirstFile(node.children);
    if (nested) {
      return nested;
    }
  }
  return null;
}

export function collectDirectoryPaths(nodes: TreeNode[], depth = 0): string[] {
  const paths: string[] = [];
  for (const node of nodes) {
    if (node.type !== "directory") {
      continue;
    }
    if (depth < 2) {
      paths.push(node.path);
    }
    paths.push(...collectDirectoryPaths(node.children, depth + 1));
  }
  return paths;
}

File: apps/api/app/analysis/tree.py
from collections.abc import Iterable
from typing import Any

from app.models import FileRecord


def build_file_tree(files: Iterable[FileRecord]) -> list[dict[str, Any]]:
    root: dict[str, Any] = {}

    for file in sorted(files, key=lambda item: item.path.lower()):
        parts = file.path.split("/")
        cursor = root
        for index, part in enumerate(parts):
            is_file = index == len(parts) - 1
            if part not in cursor:
                cursor[part] = {
                    "_type": "file" if is_file else "directory",
                    "_path": "/".join(parts[: index + 1]),
                    "_file": file if is_file else None,
                    "_children": {},
                }
            cursor = cursor[part]["_children"]

    def serialize(nodes: dict[str, Any]) -> list[dict[str, Any]]:
        output = []
        for name, node in sorted(
            nodes.items(), key=lambda item: (item[1]["_type"] == "file", item[0].lower())
        ):
            file = node["_file"]
            output.append(
                {
                    "id": file.id if file else f"dir:{node['_path']}",
                    "name": name,
                    "path": node["_path"],
                    "type": node["_type"],
                    "file_id": file.id if file else None,
                    "language": file.language if file else None,
                    "children": serialize(node["_children"]),
                }
            )
        return output

    return serialize(root)
